v2/spiffetls/dial.go

```diff
@@ -5,18 +5,29 @@ import (
 	"crypto/tls"
 
 	"github.com/spiffe/go-spiffe/v2/spiffetls/tlsconfig"
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
 )
 
 // Conn is a (m)TLS connection backed using materials obtained from the
 // Workload API.
 type Conn struct {
 	*tls.Conn
+	source *workloadapi.X509Source
 }
 
 // Close closes the connection. If the connection has been established using a
 // new X.509 source (the default behavior), that source will also be closed.
 func (c *Conn) Close() error {
-	panic("not implemented")
+	if c.source != nil {
```

Comment:
This source should only be closed if it was opened by the call to Dial (i.e. not provided via options), i.e. you need a `sourceOwned` bool or something to track ownership.

Reply:
Right, I was thinking in not storing the source pointer in the struct if it was provided by the caller. Something like:

````go
func DialWithMode(ctx context.Context, network, addr string, mode DialMode, options ...DialOption) (*Conn, error) {

			...

	// Do not store source if provided by caller
	if m.source != nil {
		source = nil
	}

	return &Conn{
		Conn:   conn,
		source: source,
	}, nil
}
````

What do you think?

Reply:
Oh, that seems fine. Maybe in that case, instead of storing the source, we should just store an `io.Closer` so its clear that the intention of holding onto it is only for the sake of closing.